File: app/src/main/java/de/westnordost/streetcomplete/quests/note_discussion/NoteDiscussionForm.java
package de.westnordost.streetcomplete.quests.note_discussion;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.graphics.drawable.RoundedBitmapDrawable;
import android.support.v4.graphics.drawable.RoundedBitmapDrawableFactory;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.io.File;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import de.westnordost.streetcomplete.Injector;
import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.OsmModule;
import de.westnordost.streetcomplete.data.osmnotes.OsmNoteQuestDao;
import de.westnordost.streetcomplete.quests.AbstractQuestAnswerFragment;
import de.westnordost.osmapi.notes.NoteComment;
import de.westnordost.streetcomplete.util.BitmapUtil;
import de.westnordost.streetcomplete.util.TextChangedWatcher;
import de.westnordost.streetcomplete.view.ListAdapter;

import static android.text.format.DateUtils.MINUTE_IN_MILLIS;

public class NoteDiscussionForm extends AbstractQuestAnswerFragment
{
	public static final String TEXT = "text";
	public static final String IMAGE_PATHS = "image_paths";

	private Bitmap anonAvatar;

	@Inject OsmNoteQuestDao noteDb;

	private EditText noteInput;
	private View buttonOk;

	@Override
	public void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		Injector.instance.getApplicationComponent().inject(this);
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
	{
		View view = super.onCreateView(inflater, container, savedInstanceState);

		View contentView = setContentView(R.layout.quest_note_discussion_content);

		View buttonPanel = setButtonsView(R.layout.quest_buttonpanel_notediscussion);
		buttonOk = buttonPanel.findViewById(R.id.buttonOk);
		buttonOk.setOnClickListener(v -> onClickOk());
		Button buttonNo = buttonPanel.findViewById(R.id.buttonNo);
		buttonNo.setOnClickListener(v -> skipQuest());

		noteInput = contentView.findViewById(R.id.noteInput);
		noteInput.addTextChangedListener(new TextChangedWatcher(this::updateOkButtonEnablement));

		buttonOtherAnswers.setVisibility(View.GONE);

		updateOkButtonEnablement();

		return view;
	}

	@Override public void onViewCreated(View view, @Nullable Bundle savedInstanceState)
	{
		super.onViewCreated(view, savedInstanceState);

		anonAvatar = BitmapUtil.createBitmapFrom(getResources().getDrawable(R.drawable.ic_osm_anon_avatar));

		inflateNoteDiscussion(noteDb.get(getQuestId()).getNote().comments);

		if(savedInstanceState == null)
		{
			getChildFragmentManager().beginTransaction().add(R.id.attachPhotoFragment, new AttachPhotoFragment()).commit();
		}
	}

	private @Nullable AttachPhotoFragment getAttachPhotoFragment()
	{
		return (AttachPhotoFragment) getChildFragmentManager().findFragmentById(R.id.attachPhotoFragment);
	}

	private void inflateNoteDiscussion(List<NoteComment> comments)
	{
		LinearLayout layout = getView().findViewById(R.id.scrollViewChild);
		RecyclerView discussionView =(RecyclerView) getLayoutInflater().inflate(R.layout.quest_note_discussion_items, layout, false);

		discussionView.setNestedScrollingEnabled(false);
		discussionView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
		discussionView.setAdapter(new NoteCommentListAdapter(comments));

		layout.addView(discussionView, 0);
	}

	private void onClickOk()
	{
		AttachPhotoFragment f = getAttachPhotoFragment();
		Bundle answer = new Bundle();
		answer.putString(TEXT, getNoteText());
		if(f != null) answer.putStringArrayList(IMAGE_PATHS, f.getImagePaths());
		applyAnswer(answer);
	}

	@Override public void onDiscard()
	{
		AttachPhotoFragment f = getAttachPhotoFragment();
		if(f != null) f.deleteImages();
	}

	private String getNoteText() { return noteInput.getText().toString().trim(); }

	@Override public boolean isRejectingClose()
	{
		AttachPhotoFragment f = getAttachPhotoFragment();
		boolean hasPhotos = f != null && !f.getImagePaths().isEmpty();
		return hasPhotos || !getNoteText().isEmpty();
	}

	private void updateOkButtonEnablement()
	{
		buttonOk.setEnabled(!getNoteText().isEmpty());
	}


	private class NoteCommentListAdapter extends ListAdapter<NoteComment>
	{
		public NoteCommentListAdapter(List<NoteComment> list) { super(list); }

		@NonNull @Override
		public ViewHolder<NoteComment> onCreateViewHolder(@NonNull ViewGroup parent, int viewType)
		{
			return new NoteCommentViewHolder(
				getLayoutInflater().inflate(R.layout.quest_note_discussion_item, parent, false));
		}
	}

	private class NoteCommentViewHolder extends ListAdapter.ViewHolder<NoteComment>
	{
		private ViewGroup commentContainer;
		private ImageView commentAvatar;
		private TextView commentText;
		private TextView commentInfo;
		private TextView commentStatusText;

		public NoteCommentViewHolder(View itemView)
		{
			super(itemView);
			commentContainer = itemView.findViewById(R.id.comment);
			commentAvatar = itemView.findViewById(R.id.comment_avatar);
			commentText = itemView.findViewById(R.id.comment_text);
			commentInfo = itemView.findViewById(R.id.comment_info);
			commentStatusText = itemView.findViewById(R.id.comment_status_text);
		}

		@Override protected void onBind(NoteComment comment)
		{
			CharSequence dateDescription = DateUtils.getRelativeTimeSpanString(
				comment.date.getTime(), new Date().getTime(), MINUTE_IN_MILLIS);

			String userName = comment.user != null
				? comment.user.displayName : getString(R.string.quest_noteDiscussion_anonymous);

			int commentActionResourceId = getNoteCommentActionResourceId(comment.action);
			if(commentActionResourceId != 0)
			{
				commentStatusText.setVisibility(View.VISIBLE);
				commentStatusText.setText(getString(commentActionResourceId, userName, dateDescription));
			} else {
				commentStatusText.setVisibility(View.GONE);
			}

			if(comment.text != null && !comment.text.isEmpty()) {
				commentContainer.setVisibility(View.VISIBLE);
				commentText.setText(comment.text);
				commentInfo.setText(getString(R.string.quest_noteDiscussion_comment2, userName, dateDescription));

				Bitmap bitmap = anonAvatar;
				if(comment.user != null)
				{
					File avatarFile = new File(OsmModule.getAvatarsCacheDirectory(getContext()) + File.separator + comment.user.id);
					if(avatarFile.exists())
					{
						bitmap = BitmapFactory.decodeFile(avatarFile.getPath());
					}
				}
				RoundedBitmapDrawable avatarDrawable = RoundedBitmapDrawableFactory.create(getResources(), bitmap);
				avatarDrawable.setCircular(true);
				commentAvatar.setImageDrawable(avatarDrawable);
			} else {
				commentContainer.setVisibility(View.GONE);
			}
		}

		private int getNoteCommentActionResourceId(NoteComment.Action action)
		{
			switch (action)
			{
				case CLOSED:		return R.string.quest_noteDiscussion_closed2;
				case REOPENED:		return R.string.quest_noteDiscussion_reopen2;
				case HIDDEN:		return R.string.quest_noteDiscussion_hide2;
				default:			return 0;
			}
		}
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/note_discussion/NoteDiscussionForm.kt
package de.westnordost.streetcomplete.quests.note_discussion

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.text.format.DateUtils
import android.text.format.DateUtils.MINUTE_IN_MILLIS
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isGone
import androidx.core.widget.doAfterTextChanged
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
import de.westnordost.streetcomplete.data.osmnotes.NoteComment
import de.westnordost.streetcomplete.data.osmnotes.edits.NoteEditAction
import de.westnordost.streetcomplete.data.osmnotes.edits.NoteEditsController
import de.westnordost.streetcomplete.data.osmnotes.edits.NotesWithEditsSource
import de.westnordost.streetcomplete.data.osmnotes.notequests.OsmNoteQuestController
import de.westnordost.streetcomplete.data.quest.OsmNoteQuestKey
import de.westnordost.streetcomplete.data.quest.QuestType
import de.westnordost.streetcomplete.data.user.User
import de.westnordost.streetcomplete.databinding.QuestNoteDiscussionContentBinding
import de.westnordost.streetcomplete.databinding.QuestNoteDiscussionItemBinding
import de.westnordost.streetcomplete.databinding.QuestNoteDiscussionItemsBinding
import de.westnordost.streetcomplete.quests.AbstractQuestForm
import de.westnordost.streetcomplete.quests.AnswerItem
import de.westnordost.streetcomplete.util.ktx.createBitmap
import de.westnordost.streetcomplete.util.ktx.nonBlankTextOrNull
import de.westnordost.streetcomplete.util.ktx.nowAsEpochMilliseconds
import de.westnordost.streetcomplete.util.ktx.viewLifecycleScope
import de.westnordost.streetcomplete.view.CircularOutlineProvider
import de.westnordost.streetcomplete.view.ListAdapter
import de.westnordost.streetcomplete.view.RoundRectOutlineProvider
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.koin.android.ext.android.inject
import org.koin.core.qualifier.named
import java.io.File

class NoteDiscussionForm : AbstractQuestForm() {

    override val contentLayoutResId = R.layout.quest_note_discussion_content
    private val binding by contentViewBinding(QuestNoteDiscussionContentBinding::bind)

    override val defaultExpanded = false

    private lateinit var anonAvatar: Bitmap

    private val noteSource: NotesWithEditsSource by inject()
    private val noteEditsController: NoteEditsController by inject()
    private val osmNoteQuestController: OsmNoteQuestController by inject()
    private val avatarsCacheDir: File by inject(named("AvatarsCacheDirectory"))

    private val attachPhotoFragment get() =
        childFragmentManager.findFragmentById(R.id.attachPhotoFragment) as? AttachPhotoFragment

    private val noteText: String? get() = binding.noteInput.nonBlankTextOrNull

    private val noteId: Long get() = (questKey as OsmNoteQuestKey).noteId

    interface Listener {
        /** Called when the user successfully answered the quest */
        fun onNoteQuestSolved(questType: QuestType, noteId: Long, position: LatLon)
        /** Called when the user did not answer the quest but also did not hide it */
        fun onNoteQuestClosed()
    }
    private val listener: Listener? get() = parentFragment as? Listener ?: activity as? Listener

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val alreadyHidden = osmNoteQuestController.getVisible(noteId) == null
        setButtonPanelAnswers(listOf(
            if (alreadyHidden) AnswerItem(R.string.short_no_answer_on_button) { closeQuest() }
            else               AnswerItem(R.string.quest_noteDiscussion_no) { hideQuest() }
        ))

        binding.noteInput.doAfterTextChanged { checkIsFormComplete() }

        anonAvatar = requireContext().getDrawable(R.drawable.ic_osm_anon_avatar)!!.createBitmap()

        viewLifecycleScope.launch {
            val comments = withContext(Dispatchers.IO) { noteSource.get(noteId) }!!.comments
            inflateNoteDiscussion(comments)
        }
    }

    private fun inflateNoteDiscussion(comments: List<NoteComment>) {
        val discussionView = QuestNoteDiscussionItemsBinding.inflate(layoutInflater, scrollViewChild, false).root

        discussionView.isNestedScrollingEnabled = false
        discussionView.layoutManager = LinearLayoutManager(
            context,
            RecyclerView.VERTICAL,
            false
        )
        discussionView.adapter = NoteCommentListAdapter(comments)

        scrollViewChild.addView(discussionView, 0)
    }

    override fun onClickOk() {
        require(noteText != null ) { "NoteQuest has been answered with an empty comment!" }
        val imagePaths = attachPhotoFragment?.imagePaths.orEmpty()
        viewLifecycleScope.launch {
            withContext(Dispatchers.IO) {
                noteEditsController.add(noteId, NoteEditAction.COMMENT, geometry.center, noteText, imagePaths)
            }
            listener?.onNoteQuestSolved(questType, noteId, geometry.center)
        }
    }

    private fun closeQuest() {
        listener?.onNoteQuestClosed()
    }

    private fun hideQuest() {
        viewLifecycleScope.launch {
            withContext(Dispatchers.IO) { osmNoteQuestController.hide(noteId) }
        }
    }

    override fun onDiscard() {
        attachPhotoFragment?.deleteImages()
    }

    override fun isRejectingClose(): Boolean =
        noteText != null || attachPhotoFragment?.imagePaths?.isNotEmpty() == true

    override fun isFormComplete(): Boolean = noteText != null

    private inner class NoteCommentListAdapter(list: List<NoteComment>) : ListAdapter<NoteComment>(list) {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<NoteComment> {
            return NoteCommentViewHolder(
                QuestNoteDiscussionItemBinding.inflate(layoutInflater, parent, false)
            )
        }
    }

    private inner class NoteCommentViewHolder(private val itemBinding: QuestNoteDiscussionItemBinding) :
        ListAdapter.ViewHolder<NoteComment>(itemBinding) {

        init {
            val cornerRadius = resources.getDimension(R.dimen.speech_bubble_rounded_corner_radius)
            val margin = resources.getDimensionPixelSize(R.dimen.horizontal_speech_bubble_margin)
            val marginStart = -resources.getDimensionPixelSize(R.dimen.quest_form_speech_bubble_top_margin)
            itemBinding.commentStatusText.outlineProvider = RoundRectOutlineProvider(cornerRadius)

            val isRTL = itemView.resources.configuration.layoutDirection == View.LAYOUT_DIRECTION_RTL
            val marginLeft = if (isRTL) 0 else marginStart
            val marginRight = if (isRTL) marginStart else 0
            itemBinding.commentBubble.outlineProvider = RoundRectOutlineProvider(
                cornerRadius, marginLeft, margin, marginRight, margin
            )
            itemBinding.commentAvatarImageContainer.outlineProvider = CircularOutlineProvider
        }

        override fun onBind(with: NoteComment) {
            val dateDescription = DateUtils.getRelativeTimeSpanString(with.timestamp, nowAsEpochMilliseconds(), MINUTE_IN_MILLIS)
            val userName = if (with.user != null) with.user.displayName else getString(R.string.quest_noteDiscussion_anonymous)

            val commentActionResourceId = with.action.actionResourceId
            val hasNoteAction = commentActionResourceId != 0
            itemBinding.commentStatusText.isGone = !hasNoteAction
            if (hasNoteAction) {
                itemBinding.commentStatusText.text = getString(commentActionResourceId, userName, dateDescription)
            }

            val hasComment = with.text?.isNotEmpty() == true
            itemBinding.commentView.isGone = !hasComment
            if (hasComment) {
                itemBinding.commentText.text = with.text
                itemBinding.commentInfoText.text = getString(R.string.quest_noteDiscussion_comment2, userName, dateDescription)

                val bitmap = with.user?.avatar ?: anonAvatar
                itemBinding.commentAvatarImage.setImageBitmap(bitmap)
            }
        }

        private val User.avatar: Bitmap? get() {
            val file = File(avatarsCacheDir.toString() + File.separator + id)
            return if (file.exists()) BitmapFactory.decodeFile(file.path) else null
        }

        private val NoteComment.Action.actionResourceId get() = when (this) {
            NoteComment.Action.CLOSED -> R.string.quest_noteDiscussion_closed2
            NoteComment.Action.REOPENED -> R.string.quest_noteDiscussion_reopen2
            NoteComment.Action.HIDDEN -> R.string.quest_noteDiscussion_hide2
            else -> 0
        }
    }
}
